controllers/podnetworkchaos/controller.go

```diff
@@ -29,76 +31,134 @@ import (
 	}
 
-	if !h.AllowHostNetworkTesting {
+	failedMessage := ""
+	observedGeneration := obj.ObjectMeta.Generation
+	defer func() {
+		if err != nil {
+			failedMessage = err.Error()
+		}
+
+		updateError := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			obj := &v1alpha1.PodNetworkChaos{}
+
+			if err := r.Client.Get(context.TODO(), req.NamespacedName, obj); err != nil {
+				r.Log.Error(err, "unable to get chaos")
+				return err
+			}
+
+			obj.Status.FailedMessage = failedMessage
+			obj.Status.ObservedGeneration = observedGeneration
+
+			return r.Client.Status().Update(context.TODO(), obj)
+		})
+
+		if updateError != nil {
+			r.Log.Error(updateError, "fail to update")
+			r.Recorder.Eventf(obj, "Normal", "Failed", "Failed to update status: %s", updateError.Error())
+		}
+	}()
+
+	if !r.AllowHostNetworkTesting {
 		if pod.Spec.HostNetwork {
-			err := errors.Errorf("it's dangerous to inject network chaos on a pod(%s/%s) with `hostNetwork`", pod.Namespace, pod.Name)
-			return err
+			err = errors.Errorf("It's dangerous to inject network chaos on a pod(%s/%s) with `hostNetwork`", pod.Namespace, pod.Name)
+			r.Log.Error(err, "fail to inject network chaos")
+			r.Recorder.Event(obj, "Warning", "Failed", err.Error())
+			return ctrl.Result{}, nil
 		}
 	}
 
-	err = h.SetIPSets(ctx, pod, chaos)
+	err = r.SetIPSets(ctx, pod, obj)
 	if err != nil {
-		return err
+		r.Log.Error(err, "fail to set ipsets")
+		r.Recorder.Eventf(obj, "Warning", "Failed", "Fail to set ipsets: %s", err.Error())
+		return ctrl.Result{}, nil
 	}
 
-	err = h.SetIptables(ctx, pod, chaos)
+	err = r.SetIptables(ctx, pod, obj)
 	if err != nil {
-		return err
+		r.Log.Error(err, "fail to set iptables")
+		r.Recorder.Eventf(obj, "Warning", "Failed", "Fail to set iptables: %s", err.Error())
+		return ctrl.Result{}, nil
 	}
 
-	err = h.SetTcs(ctx, pod, chaos)
+	err = r.SetTcs(ctx, pod, obj)
 	if err != nil {
-		return err
+		r.Recorder.Eventf(obj, "Warning", "Failed", "Fail to set tc: %s", err.Error())
+		return ctrl.Result{}, nil
 	}
+	r.Recorder.Event(obj, "Normal", "Updated", "Synchronize podnetworkchaos successfully")
 
-	return nil
+	return ctrl.Result{}, nil
 }
 
 // SetIPSets sets ipset on pod
-func (h *Handler) SetIPSets(ctx context.Context, pod *corev1.Pod, chaos *v1alpha1.PodNetworkChaos) error {
+func (r *Reconciler) SetIPSets(ctx context.Context, pod *corev1.Pod, chaos *v1alpha1.PodNetworkChaos) error {
 	ipsets := []*pb.IPSet{}
 	for _, ipset := range chaos.Spec.IPSets {
 		ipsets = append(ipsets, &pb.IPSet{
 			Name:  ipset.Name,
 			Cidrs: ipset.Cidrs,
 		})
 	}
-	return ipset.FlushIPSets(ctx, h.Client, pod, ipsets)
+	return ipset.FlushIPSets(ctx, r.Client, pod, ipsets)
 }
 
 // SetIptables sets iptables on pod
-func (h *Handler) SetIptables(ctx context.Context, pod *corev1.Pod, chaos *v1alpha1.PodNetworkChaos) error {
+func (r *Reconciler) SetIptables(ctx context.Context, pod *corev1.Pod, chaos *v1alpha1.PodNetworkChaos) error {
```

Comment:
It seems we can manage the iptables rules by PodNetworkChaos instead of PodHttpChaos itself. ( However, we need more features here in that case.

Reply:
Yes. This structure design is very limited. Have you tested, whether `podhttpchaos` works well with `podnetworkchaos` (except the double delay problem)?

Reply:
Not yet, I will test it in implementing the `httpchaos`.